src/test_record.rs

```diff
@@ -0,0 +1,257 @@
+//! Support for recording network activity for test playback.
+//!
+//! See `testsuite.rs` for more information on test recording.
+
+use crate::EventName;
+use anyhow::Context;
+use anyhow::Result;
+use reqwest::{Request, StatusCode};
+use serde::{Deserialize, Serialize};
+use std::fs::{self, File};
+use std::io::BufWriter;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::atomic::{AtomicU32, Ordering};
+use tracing::{error, info, warn};
+
+/// A representation of the recording of activity of triagebot.
+///
+/// Activities are stored as JSON on disk. The test framework injects the
+/// `Webhook` and then checks for the appropriate requests and sends the
+/// recorded responses.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(tag = "kind")]
+pub enum Activity {
+    /// A webhook event received from GitHub.
+    Webhook {
+        webhook_event: String,
+        payload: serde_json::Value,
+    },
+    /// An outgoing request to api.github.com, and its response.
+    ApiRequest {
+        method: String,
+        path: String,
+        query: Option<String>,
+        request_body: String,
+        response_code: u16,
+        response_body: serde_json::Value,
+    },
+    /// An outgoing request to raw.githubusercontent.com, and its response.
+    RawRequest {
+        path: String,
+        query: Option<String>,
+        response_code: u16,
+        response_body: String,
+    },
+    /// Sent by the mock HTTP server to the test framework when it detects
+    /// something is wrong.
+    Error { message: String },
+    /// Sent by the mock HTTP server to the test framework to tell it that all
+    /// activities have been processed.
+    Finished,
+}
+
+/// Information about an HTTP request that is captured before sending.
+///
+/// This is needed to avoid messing with cloning the Request.
+pub struct RequestInfo {
+    /// If this is `true`, then it is for raw.githubusercontent.com.
+    /// If `false`, then it is for api.github.com.
+    is_raw: bool,
+    method: String,
+    path: String,
+    query: Option<String>,
+    body: String,
+}
+
+/// Returns whether or not TRIAGEBOT_TEST_RECORD has been set to enable
+/// recording.
+pub fn is_recording() -> bool {
+    record_dir().is_some()
+}
+
+/// The directory where the JSON recordings should be stored.
+///
+/// Returns `None` if recording is disabled.
+fn record_dir() -> Option<PathBuf> {
+    let Some(test_record) = std::env::var_os("TRIAGEBOT_TEST_RECORD") else { return None };
```

Comment:
I'm not entirely clear on this suggestion. Each time an event comes in, it needs to record it to disk. In order for it to know where to record it, it needs to read the `TRIAGEBOT_TEST_RECORD`. It can't be checked in one place. Unless this is perhaps suggesting to store the value somewhere and read that directly? It seems like that would make it a bit more difficult to use since there would need to be some global state tracked somewhere.

I added a brief mention of recording in README.md